### PR TITLE
[Sema] Warn on unsound pointer conversions

### PR DESCRIPTION
**Pull request by hamishknight**
In this PR is the logic to emit a warning on unsound array-to-pointer, string-to-pointer, and inout-to-pointer conversions. This is done through the introduction of an underscored attribute `@_nonEphemeral` which denotes a pointer parameter which cannot accept pointer conversions that are only valid for the duration of the call they're passed to.

This includes all array-to-pointer and string-to-pointer conversions, and most inout-to-pointer conversions with the exception of:
- Global and static variables
- Stored properties on such variables that are accessed directly (e.g excluding ones with property observers, ones on non-final classes and resilient types in general)
- Force unwraps of any of the above

Various pointer parameters in the stdlib have been marked as `@_nonEphemeral`, primarily pointer parameters for initialisers on pointer types. In addition, this PR infers `@_nonEphemeral` on pointer parameters for memberwise initialisers and enum case constructors.

Essentially, this is the warning logic of #20070 plus:
- ~~The allowing of stored properties on final class global/static vars to be treated as non-ephemeral.~~
- An additional note that suggests that the user use `withUnsafe[Mutable][Bytes/Pointer]`, or for arrays, `withUnsafe[Mutable][Bytes/BufferPointer]`, or for strings, `withCString`.
- The inferring of `@_nonEphemeral` for enum case constructors.
- More tests.

If possible, I'd like to get this into Swift 5, allowing us to transition to an error for Swift 6. I do appreciate how late we are into the release cycle for such a (relatively big) change though.

I'm unsure if this PR requires Swift evolution discussion – I'm happy to start up a thread if it's felt that it does.

Resolves [SR-2790].

  [SR-2790]: https://bugs.swift.org/browse/SR-2790

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -365,6 +365,35 @@ ERROR(cannot_convert_argument_value_generic,none,
       "cannot convert value of type %0 (%1) to expected argument type %2 (%3)",
       (Type, StringRef, Type, StringRef))
 
+ERROR(cannot_convert_non_ephemeral,none,
+      "temporary pointer argument of type %0 cannot be passed to parameter "
+      "expecting pointer that outlives the duration of the call", (Type))
+WARNING(cannot_convert_non_ephemeral_warning,none,
+        "passing temporary pointer argument of type %0 to parameter expecting "
+        "pointer that outlives the duration of the call leads to undefined "
+        "behaviour; this will be an error in a future release", (Type))
```
Comment (hamishknight) [inline include/swift/AST/DiagnosticsSema.def]: This diagnostic feels a bit too wordy, but I can't think of how to improve it – any ideas @jrose-apple?

Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: Hm, well, we could remove that "u" from "behavior". :-)

> warning: temporary pointer will be destroyed after call to %0

Okay, that sounds sufficiently scary, but the problem is that the user probably doesn't think they've written a pointer. They wrote an inout expression or an array or string.

> error: cannot pass array to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot pass string to %0; argument %1 must be a pointer that outlives the call to %0
> error: cannot use inout expression here; argument %1 must be a pointer that outlives the call to %0

How does that sound? And then we get the warning form for pre-Swift-5 (which we maybe take away completely in the future):

> warning: passing array to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: passing string to %0, but argument %1 should be a pointer that outlives the call to %0
> warning: inout expression creates a temporary pointer, but argument %1 should be a pointer that outlives the call to %0

I don't love this but it's a little less wordy?

Comment (hamishknight) [reply include/swift/AST/DiagnosticsSema.def]: Oops, need to start using a US spell checker :) Thanks for the suggestions, I like them!

IMO it would be nice to use the argument type of the array or string in the diagnostic message, and avoid repeating the name of the callee – so how about the following?

> error: cannot pass %0 to parameter; argument must a pointer that outlives the call%select{| to %2}1

> warning: passing %0 to parameter, but argument should be a pointer that outlives the call%select{| to %2}1; this will be an error in a future release

> error: cannot use inout expression here; argument must a pointer that outlives the call%select{| to %1}0

> warning: inout expression creates a temporary pointer, but argument should be a pointer that outlives the call%select{| to %1}0; this will be an error in a future release

For now, I've omitted the placeholder for "argument %1", as I wasn't too sure what it would be filled in with – would it be the argument position or the text of the argument expr?

I've gone ahead and pushed a commit with the message changes so far.

Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: Typo: "must a pointer".

The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me. The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

**Review by atrick — COMMENTED** [review]
I'll let @jrose-apple approve this because I don't know Sema well enough to determine if you're missing anything. But I did review the code and tests, and they looks astonishingly thorough to me.

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/DiagnosticsSema.def]: > Typo: "must a pointer"

Oops, good catch!

> The reason I had the name of the callee twice was that "to parameter" on its own sounds weird to me.

Interesting! "to parameter" sounds natural to me – nonetheless I'm happy to change it. Though for cases where we don't have a callee decl, should it be "to parameter" or "to call"? The former sounds more natural to me, but the latter is more consistent with the message when we do have a callee.

> The "argument %1" part was the position in my head, to make it slightly easier to tell what the diagnostic is referring to when just looking at the line and not the column or highlight.

Thanks for clarifying, I'll go ahead and make that change.

> I feel like the "this will be an error in a future release" isn't really adding that much value either, especially for a message that's already fairly long.

Fair enough – I added it because I felt the diagnostic wasn't quite scary enough without it, but given the length of the message, I agree it probably doesn't quite pull its weight.

File: test/Sema/diag_non_ephemeral_swift4.swift [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+takesOptMutableRaw(&value)
+takesOptMutableRaw(&str)
+
+extension C {
+  static func foo() {
+    takesRaw(&staticStoredProperty)
+  }
+}
+
+// We can also project stable and, in the case of final classes, semi-stable pointer values
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply test/Sema/diag_non_ephemeral_swift4.swift]: Ah sorry, I should have expanded on that a bit more – what I meant was "stable until the object is deallocated". I'll edit the comment.

Thanks for looking over the PR!

File: lib/Sema/ConstraintSystem.cpp [hunk 3]
```diff
@@ -2356,3 +2356,110 @@ Expr *constraints::simplifyLocatorToAnchor(ConstraintSystem &cs,
...
+  case ConversionRestrictionKind::ClassMetatypeToAnyObject:
+  case ConversionRestrictionKind::ExistentialMetatypeToAnyObject:
+  case ConversionRestrictionKind::ProtocolMetatypeToProtocolClass:
+  case ConversionRestrictionKind::PointerToPointer:
+  case ConversionRestrictionKind::ArrayUpcast:
+  case ConversionRestrictionKind::DictionaryUpcast:
+  case ConversionRestrictionKind::SetUpcast:
+  case ConversionRestrictionKind::HashableToAnyHashable:
+  case ConversionRestrictionKind::CFTollFreeBridgeToObjC:
+  case ConversionRestrictionKind::ObjCTollFreeBridgeToCF:
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Sema/ConstraintSystem.cpp]: Should this be llvm_unreachable, then?

File: lib/Sema/CSDiag.cpp [hunk 4]
```diff
@@ -8403,3 +8412,83 @@ bool swift::diagnoseBaseUnwrapForMemberAccess(Expr *baseExpr, Type baseType,
...
+    } else {
+      return false;
+    }
+
+    auto diagID = downgradeToWarning
+                      ? diag::cannot_construct_dangling_pointer_warning
+                      : diag::cannot_construct_dangling_pointer;
+
+    diags.diagnose(anchor->getLoc(), diagID, instanceTy, constructorKind)
+        .highlight(anchor->getSourceRange());
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Is the anchor expression always the correct expression? What if the construction is nested inside a larger expression?

File: lib/Sema/CSDiag.cpp [hunk 5]
```diff
@@ -161,6 +161,15 @@ void constraints::simplifyLocator(Expr *&anchor,
...
       }
 
+      if (auto *subscriptExpr = dyn_cast<SubscriptExpr>(anchor)) {
+        targetAnchor = nullptr;
+        targetPath.clear();
+
+        anchor = subscriptExpr->getIndex();
+        path = path.slice(1);
+        continue;
+      }
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: What else is this going to affect?

File: lib/Sema/TypeCheckPattern.cpp [hunk 6]
```diff
@@ -838,6 +838,16 @@ bool TypeChecker::typeCheckParameterList(ParameterList *PL,
...
     if (!hadError) {
+      // If we're in an enum case, then infer @_nonEphemeral for pointer
+      // payloads.
+      if (options.getBaseContext() == TypeResolverContext::EnumElementDecl) {
+        auto unwrappedType = type;
+        if (auto objectType = type->getOptionalObjectType())
+          unwrappedType = objectType;
+        if (unwrappedType->getAnyPointerElementType())
+          param->setNonEphemeral();
+      }
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckPattern.cpp]: Do we also want this for stored properties?

File: lib/Sema/ConstraintSystem.cpp [hunk 7]
```diff
@@ -2417,14 +2404,22 @@ bool ConstraintSystem::isConversionNonEphemeral(
         continue;
       }
 
+      // Look through load exprs. These can occur when projecting through a
+      // property on a class, which we want to support if the class is
+      // non-final.
```
Comment (jrose-apple) [inline lib/Sema/ConstraintSystem.cpp]: I don't think this is the correct condition. What case is this trying to handle?

File: lib/Sema/CSApply.cpp [hunk 8]
```diff
@@ -5475,22 +5482,31 @@ static void checkNonEphemeralArgumentConversion(
...
-  if (conversion && !cs.isConversionNonEphemeral(*conversion, argLocator))
+  if (conversion && !cs.isConversionNonEphemeral(*conversion, argLocator)) {
+    auto lastPathElt = argLocator.last();
+    assert(lastPathElt &&
+           lastPathElt->getKind() == ConstraintLocator::ApplyArgToParam);
+    auto argIdx = lastPathElt->getValue();
     diagnoseIllegalEphemeralConversion(
-        cs.getASTContext(), *conversion, argExpr, argType, param.getPlainType(),
-        callee, fnType, argLocator.getAnchor(),
+        cs.getASTContext(), *conversion, argIdx, subExpr, argType,
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Now that you're passing `subExpr`, I don't think you need to pass `argType`.

File: test/Sema/diag_non_ephemeral_swift4.swift [hunk 9]
```diff
@@ -21,46 +21,46 @@ var optionalArr: [Int8]?
 // We cannot use array-to-pointer and string-to-pointer conversions with
 // non-ephemeral parameters.
 
-takesMutableRaw(&arr, 5) // expected-warning {{passing temporary pointer argument of type 'UnsafeMutableRawPointer' to parameter expecting pointer that outlives the duration of the call leads to undefined behaviour; this will be an error in a future release}}
-// expected-note@-1 {{implicit argument conversion from '[Int8]' to 'UnsafeMutableRawPointer' produces a pointer valid only for the duration of the call}}
+takesMutableRaw(&arr, 5) // expected-warning {{inout expression creates a temporary pointer, but argument #1 should be a pointer that outlives the call to 'takesMutableRaw'}}
```
Comment (jrose-apple) [inline test/Sema/diag_non_ephemeral_swift4.swift]: It's weird to have a diag_non_ephemeral_swift4.swift without the corresponding diag_non_ephemeral_swift5.swift.

**Review by jrose-apple — COMMENTED** [review]
There's a lot here so I might have still missed something, but this is what I noticed reading through the whole thing. @xedin or @rudkx should probably comment on the ConstraintSystem parts.

File: lib/Sema/ConstraintSystem.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply lib/Sema/ConstraintSystem.cpp]: In this patch it could be, although IMO it should it be reasonable for the function to take an arbitrary conversion and answer the question "is this something that we could pass to a `@_nonEphemeral` parameter?" – which is true for the rest of these conversions, as they're not things that `@_nonEphemeral` restricts.

File: test/attr/attributes.swift [hunk 10]
```diff
@@ -273,3 +273,29 @@ class SILStored {
...
+  // FIXME(SR-9178): We shouldn't produce a duplicate diagnostic here.
+  func baz<U>(_ x: @_nonEphemeral U) {} // expected-error 2{{@_nonEphemeral attribute currently only applies to pointer types}}
+
+  func qux(_ x: @_nonEphemeral UnsafeMutableRawPointer) {}
+  func quux(_ x: @_nonEphemeral UnsafeMutablePointer<Int>?) {}
+}
+
+protocol P {}
+extension P {
+  // Allow @_nonEphemeral on the protocol Self type, as the protocol could be adopted by a pointer type.
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply test/attr/attributes.swift]: You're right, it's not technically required, but it's just something I thought would be nice. It's not entirely foolproof either, as you could mark a protocol `Self` parameter as `@_nonEphemeral` and then have the protocol adopted by a non-pointer type. I'm happy to rip out the check if you want.

File: lib/Sema/CSDiag.cpp (hunk 4 above)
Comment (hamishknight) [reply lib/Sema/CSDiag.cpp]: As I understand it, unsimplified argument locators for an apply/subscript are always anchored on the apply/subscript expr that they correspond to.

This is how they're generated in CSApply and CSGen at least:

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSGen.cpp#L2561-L2564

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSApply.cpp#L3085-L3089

File: lib/Sema/TypeCheckPattern.cpp (hunk 6 above)
Comment (hamishknight) [reply lib/Sema/TypeCheckPattern.cpp]: Yes, and we're inferring `@_nonEphemeral` for parameters of memberwise initialisers – did you have another case in mind?

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (hamishknight) [reply lib/Sema/ConstraintSystem.cpp]: This is handling the following (which we want to allow):

```swift
final class C {
  var str = ""
}

struct S {
  var c = C()
}

var s = S()
let ptr = UnsafeMutablePointer(&s.c.str)
```

The base `s` is an lvalue, but we want an rvalue `C`, so a `LoadExpr` is inserted before then performing the member access for `str`.

File: lib/Sema/CSApply.cpp (hunk 8 above)
Comment (hamishknight) [reply lib/Sema/CSApply.cpp]: `diagnoseIllegalEphemeralConversion` will eventually also want to be able to take expressions that haven't had solutions applied to them, so I *believe* the type needs to be passed separately (at least, I'm not sure whether it's safe to rely on CSApply giving us back the same `InOutExpr` such that we can query the type from the constraint system).

File: test/Sema/diag_non_ephemeral_swift4.swift (hunk 9 above)
Comment (hamishknight) [reply test/Sema/diag_non_ephemeral_swift4.swift]: Yeah, that's a holdover from when I had the logic combined for a warning in Swift 4 mode and error for Swift 5 mode. I'll rename it to `diag_non_ephemeral_warning`.

Thanks for the review!

File: lib/Sema/CSDiag.cpp (hunk 5 above)
Comment (hamishknight) [reply lib/Sema/CSDiag.cpp]: From what I could tell, the majority of usages of `simplifyLocator` occur within CSDiag, and the usages outside of CSDiag appear to revolve around locators to decls, which this shouldn't affect.

So as far as I can tell, the worst-case impact would be a slight change in diagnostics (but even then, it should be a change for the better, as we're now correctly simplifying a subscript argument locator to the argument expr).

That being said, looking at it again, it appears this logic isn't necessary for this patch, so I'll remove it to minimise risk.

File: test/attr/attributes.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply test/attr/attributes.swift]: We need this for the two initialisers defined on the stdlib's `_Pointer` protocol, which take `Self(?)` parameters and can accept X-to-pointer conversions:

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/stdlib/public/core/Pointer.swift#L79-L95

[the remark this answers is not in the document]
Comment (jrose-apple) [reply test/attr/attributes.swift]: Ugh. Okay.

File: lib/Sema/TypeCheckPattern.cpp (hunk 6 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckPattern.cpp]: I meant setters, sorry.

Comment (hamishknight) [reply lib/Sema/TypeCheckPattern.cpp]: Ah, I see! Though I can't immediately think of a case where you can use X-to-pointer conversions in conjunction with setters, AFAIK we only allow them for arguments of applies/subscripts.

**Comment by xedin** [comment]
I'll try to take a look at Sema changes tomorrow.

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (jrose-apple) [reply lib/Sema/ConstraintSystem.cpp]: I don't think we want to allow this. Basically if you load from a class, you don't care what the base was anymore, and then either we allow forming pointers to instance variables or we don't. Being a global doesn't matter at that point. Right, @atrick?

File: lib/Sema/TypeCheckPattern.cpp (hunk 6 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckPattern.cpp]: Ah, I didn't realize we didn't perform those conversions for assignments! Okay, good.

```swift
struct Test {
  var ptr: UnsafeRawPointer?
}

var x = Test()
x.ptr = "abc" // already an error
```

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: `let ptr = UnsafeMutablePointer(&s.c.str)`

If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Sorry if I missed that in my review.

Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: Looking through `diag_non_ephemeral_warning.swift` again, I see these two cases that still don't produce warnings, but probably should:
```
takesRaw(&topLevelFinalC.storedProperty)
takesRaw(&topLevelFragileFinalC.storedProperty)
```

Comment (jrose-apple) [reply lib/Sema/ConstraintSystem.cpp]: I think we probably do want to make that guarantee, but even without that differentiating between global `let` and global `var` seems way too subtle.

Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: We do support `realloc` on uniquely reference objects.

Comment (hamishknight) [reply lib/Sema/ConstraintSystem.cpp]: @jrose-apple As I understand it, the base still matters for class instances. For example, we don't want to allow users to form a pointer to an instance variable on a class-typed local variable, as we don't guarantee it will remain initialised until the end of the stack frame.

Comment (hamishknight) [reply lib/Sema/ConstraintSystem.cpp]: > `let ptr = UnsafeMutablePointer(&s.c.str)`
>
> If `s` is a `var`, then there's there's nothing keeping `c` alive. So escaping the addres of `c.str` seems pretty bad.

@atrick Should `c` not be guaranteed to be kept alive as long as `s.c` isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

> If `s` is a top-level or global `let` then `c` stays alive. However, technically we don't guarantee that a Swift class object has a persistent address (it may need to for satisfying some runtime constraint, but the language doesn't guarantee it). So I don't think we should allow that either.

Ah, okay then. I think for now at least I'll revert b6e1e996bd3c99e70c7c344d4c2ea5010cea4eb9, with the view of possibly reinstating it if we ever officially make such a guarantee. I somewhat doubt many (if any) people are using this pattern anyway – there was no sign of it in the source compat suite.

Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: > @atrick Should c not be guaranteed to be kept alive as long as s.c isn't changed to a different reference? IMO this would be reasonable constraint to place on the lifetime of the pointer created from the inout-to-pointer conversion.

Yes. If `s` is a `let`, then we have that guarantee. And on second thought, I don't think it will ever be possible to reallocate an object that has a `let` reference. But I think @jrose-apple and I agree that it's better to have a simple rule that disallows non-ephemeral pointers to instance variables.

Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: > Ah, okay then. I think for now at least I'll revert b6e1e99,

Wait a sec, that commit also allows non-ephemeral pointers to `static` variables. I think that's ok under the circumstances that you've identified.

**Review by xedin — CHANGES_REQUESTED** [review]

File: include/swift/AST/Decl.h [hunk 11]
```diff
@@ -4854,11 +4862,15 @@ class ParamDecl : public VarDecl {
     StringRef StringRepresentation;
   };
 
-  /// The default value, if any, along with whether this is varargs.
-  llvm::PointerIntPair<StoredDefaultArgument *, 1> DefaultValueAndIsVariadic;
+  enum class Flags {
```
[the remark this answers is not in the document]
Comment (hamishknight) [reply include/swift/AST/Decl.h]: Sure thing! I've opened a separate PR: https://github.com/apple/swift/pull/20663

File: lib/Sema/ConstraintSystem.cpp (hunk 7 above)
Comment (atrick) [reply lib/Sema/ConstraintSystem.cpp]: > Does this include non-ephemeral pointers to stored instance properties on global/static variable structures?

No, stored properties on structs can be pointed to if the struct itself is global/static. We want to disallow pointing to stored properties inside class instances, which I think we were calling "instance variables".

File: lib/Sema/CSApply.cpp [hunk 12]
```diff
@@ -5454,6 +5459,59 @@ static bool hasCurriedSelf(ConstraintSystem &cs, ConcreteDeclRef callee,
   return false;
 }
 
+static void checkNonEphemeralArgumentConversion(
```
[the remark this answers is not in the document]
Comment (xedin) [reply lib/Sema/CSApply.cpp]: I think it would make most sense to make it a`TypeMatchFlag`. Maybe some constraints should just carry type match options which could be deduced by `matchCallArguments`? Or otherwise it has to be a "constraint restriction"...

**Comment by hamishknight** [comment]
Sorry for taking so long to get back to this! I have opened an updated PR here: https://github.com/apple/swift/pull/27695.
